blog/2024-07-10-uppy-4.0.md

```diff
@@ -0,0 +1,421 @@
+their tails in delight. So, sit, stay, and let’s dig into what makes Uppy 4.0
+the goodest of good boys in file uploading!
+
+<!--truncate-->
+
+## Migration guide
+
+This post goes into the most exciting new features of Uppy 4.0. We have an
+accompanying [migration guide](/docs/guides/migration-guides/) to help you
+transition to 4.0.
+
+## TypeScript rewrite
+
+In the year 2024 people expect excellent types from their libraries. We used to
+author types separately by hand but they were often inconsistent or incomplete.
+Now Uppy has been completely rewritten in TypeScript!
+
+From now on you’ll be in safe hands when working with Uppy, whether it’s setting
+the right options, building plugins, or listening to events.
+
+```ts
+import Uppy from '@uppy/core';
+
+const uppy = new Uppy();
+
+// Event name autocompletion and inferred argument types
+uppy.on('file-added', (file) => {
+	console.log(file);
+});
+```
+
+One important thing to note are the new generics on `@uppy/core`.
+
+<!-- eslint-disable @typescript-eslint/no-non-null-assertion -->
+
+```ts
+import Uppy from '@uppy/core';
+// xhr-upload is for uploading to your own backend.
+import XHRUpload from '@uppy/xhr-upload';
+
+// Your own metadata on files
+type Meta = { myCustomMetadata: string };
+// The response from your server
+type Body = { someThingMyBackendReturns: string };
+
+const uppy = new Uppy<Meta, Body>().use(XHRUpload, {
+	endpoint: '/upload',
+});
+
+const id = uppy.addFile({
+	name: 'example.jpg',
+	data: new Blob(),
+	meta: { myCustomMetadata: 'foo' },
+});
+
+// This is now typed
+const { myCustomMetadata } = uppy.getFile(id).meta;
+
+await uppy.upload();
+
+// This is strictly typed too
+const { someThingMyBackendReturns } = uppy.getFile(id).response.body!;
+```
+
+Happy inferring!
+
+## Merging the two AWS S3 plugins
+
+We used to two separate plugins for uploading to S3 (and S3-compatible
+services): `@uppy/aws-s3` and `@uppy/aws-s3-multpart`. They have different use
+cases. The advantages of multipart uploads are:
+
+- Improved throughput – You can upload parts in parallel to improve throughput.
+- Quick recovery from any network issues – Smaller part size minimizes the
+  impact of restarting a failed upload due to a network error.
+- Pause and resume object uploads – You can upload object parts over time. After
+  you initiate a multipart upload, there is no expiry; you must explicitly
+  complete or stop the multipart upload.
+- Begin an upload before you know the final object size – You can upload an
+  object as you are creating it.
+
+However, the downside is request overhead, as it needs to do creation, signing,
+and completion requests besides the upload requests. For example, if you are
+uploading files that are only a couple kilobytes with a 100ms roundtrip latency,
+you are spending 400ms on overhead and only a few milliseconds on uploading.
+This really adds up if you upload a lot of small files.
```

Comment:
nit: I wonder if we'd want to present the info in a table that shows (one column for advantages, one column for disadvantages)

Reply:
I think it's fine like this. We're also only mentioning one downside so it would be a very uneven table